[css-anchor-position-1] Fix intrinsic layout of anchor-positioned element after visibility change

Scony opened the pull request:
#### 8a011645e7ae8104a27f7a3401668084311e7e63
<pre>
[css-anchor-position-1] Fix intrinsic layout of anchor-positioned element after visibility change
<a href="https://bugs.webkit.org/show_bug.cgi?id=281728">https://bugs.webkit.org/show_bug.cgi?id=281728</a>

Reviewed by Antti Koivisto.

Since style resolution is not being done in one pass when anchoring is used, block elements hold its children resolution until resolving anchors is done.
In that case - when it finally happens - the resolution of block element children is not enforced (as it would be during 1st pass) and the children are
not willing to &quot;force update&quot; and create renderers in the process. The net outcome is that children of elements that became visible again, are not rendered.

This change fixes the above issue by extending one of the conditions that leads to creation of renderers when the resolution is deferred due to anchoring.

* LayoutTests/TestExpectations:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::determineResolutionType):

Canonical link: <a href="https://commits.webkit.org/285486@main">https://commits.webkit.org/285486@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5aa9afe9fc7a68894ea3793b3fbd0a664bf10835

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72827 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52252 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25626 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77020 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24061 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74942 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60057 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23877 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57255 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15750 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75894 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47239 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62678 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37684 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43886 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20143 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22390 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65739 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20502 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78696 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/17073 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19644 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65707 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/17121 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62685 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64983 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16060 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13294 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6947 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/48050 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2837 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/49117 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50412 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48862 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -978,7 +978,7 @@ auto TreeResolver::determineResolutionType(const Element& element, const RenderS
         return ResolutionType::Full;
     case DescendantsToResolve::ChildrenWithExplicitInherit:
-        if (existingStyle && existingStyle->hasExplicitlyInheritedProperties())
+        if (!existingStyle || (existingStyle && existingStyle->hasExplicitlyInheritedProperties()))
```
The fix seems logically correct (if we don't have a style we need to compute it). Can you tell why this hasn't caused trouble elsewhere and why it does cause trouble with anchor positioning?

You can just write

`if (!existingStyle || existingStyle->hasExplicitlyInheritedProperties())`

Scony replied:
Good point.

As for why it does work without anchor positioning - in that scenarios the resolution of children of element that became visible again is being done altogether **always**.  In that case, in `WebCore::Style::TreeResolver::computeDescendantsToResolve()` we're hitting `case Change::Renderer:` that forces full resolution (yields ` DescendantsToResolve::All`). Moreover, further down the road in `TreeResolver::determineResolutionType()` we have `combinedValidity` equal to `WebCore::Style::Validity::SubtreeInvalid` (for entire subtree) that leads to full resolution of all subtree elements.  With anchor-positioning, if the element that become visible again is anchor positioned, at the point when `update.change` resolves to `Change::Renderer` only the original element creates renderer - it's children have to wait for anchor-positioning resolution. When it finally is done (few passes after) the `update.change` resolves to `DescendantsToResolve::ChildrenWithExplicitInherit` which then leads to no resolution in `TreeResolver::determineResolutionType()` as such corner-case was never happening before. The fix proposed in this PR just handles such unhanded case.

Scony committed:
[css-anchor-position-1] Fix intrinsic layout of anchor-positioned element after visibility change
https://bugs.webkit.org/show_bug.cgi?id=281728

Reviewed by Antti Koivisto.

Since style resolution is not being done in one pass when anchoring is used, block elements hold its children resolution until resolving anchors is done.
In that case - when it finally happens - the resolution of block element children is not enforced (as it would be during 1st pass) and the children are
not willing to "force update" and create renderers in the process. The net outcome is that children of elements that became visible again, are not rendered.

This change fixes the above issue by extending one of the conditions that leads to creation of renderers when the resolution is deferred due to anchoring.

* LayoutTests/TestExpectations:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::determineResolutionType):

Canonical link: https://commits.webkit.org/285486@main
